### PR TITLE
ISPN-8517 Lazily ressurect ice fromMemory

### PR DESCRIPTION
**Pull request by wburns**
* Reuse address pointer for compute remove
* Don't resurrect value if not needed

https://issues.jboss.org/browse/ISPN-8517
https://issues.jboss.org/browse/ISPN-8500

This issue now also includes objects that are expired are not read.

### CODE REVIEW

**Comment by wburns** [comment]
This should cherry pick as is into 9.1.x as well.

**Comment by wburns** [comment]
Test failure is unrelated.

**Comment by wburns** [comment]
Added in changes from #5567

**Comment by wburns** [comment]
Looks like CI failures were from the random tx failures in #5591. I have rebased and letting CI run again.

File: core/src/main/java/org/infinispan/container/offheap/OffHeapEntryFactoryImpl.java [hunk 1]
```diff
@@ -359,6 +360,78 @@ public boolean equalsKey(long address, WrappedBytes wrappedBytes) {
       return true;
    }
 
+   /**
+    * Returns whether entry is expired or not.
+    * @param address the address of the entry's key to check
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/container/offheap/OffHeapEntryFactoryImpl.java]: Isn't it the address of the entry?

File: core/src/main/java/org/infinispan/container/offheap/OffHeapEntryFactoryImpl.java [hunk 2]
```diff
@@ -359,6 +360,78 @@ public boolean equalsKey(long address, WrappedBytes wrappedBytes) {
       return true;
    }
 
+   /**
+    * Returns whether entry is expired or not.
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/container/offheap/OffHeapEntryFactoryImpl.java]: Seems redundant.

File: core/src/test/java/org/infinispan/expiration/impl/ExpirationFunctionalTest.java [hunk 3]
```diff
@@ -34,15 +47,25 @@ protected EmbeddedCacheManager createCacheManager(ConfigurationBuilder builder)
...
+            .expiration().disableReaper()
+            .memory().storageType(storage);
    }
 
    protected void afterCacheCreated(EmbeddedCacheManager cm) {
 
    }
 
-   public void testSimpleExpirationLifespan() throws Exception {
+   public ExpirationFunctionalTest withStorage(StorageType storage) {
```
Comment (danberindei) [inline core/src/test/java/org/infinispan/expiration/impl/ExpirationFunctionalTest.java]: Needs `parameters()`.

File: core/src/test/java/org/infinispan/expiration/impl/ExpirationFunctionalTest.java [hunk 4]
```diff
@@ -51,11 +74,58 @@ public void testSimpleExpirationLifespan() throws Exception {
...
+   }
+
+   public void testExpirationMaxIdleInExec() throws Exception {
+      for (int i = 0; i < SIZE; i++) {
+         cache.put("key-" + i, "value-" + i,-1, null, 1, TimeUnit.MILLISECONDS);
+      }
+      timeService.advance(2);
+
+      for (int i = 0; i < SIZE; i++) {
+         cache.getAdvancedCache().getDataContainer().executeTask(KeyFilter.ACCEPT_ALL_FILTER, (k, ice) -> { throw new RuntimeException("No task should be executed on expired entry"); });
```
Comment (danberindei) [inline core/src/test/java/org/infinispan/expiration/impl/ExpirationFunctionalTest.java]: The lambda could be on its own line.

**Comment by wburns** [comment]
Okay this has been updated and also fixes the REST failures on master.

**Pull request merged**